src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1523,6 +1523,134 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
+    Label* larr[] = { &case0, &case1, &case2, &case3, &case4, &case5, &case6, &case7 };
+    for (int i = 0; i < 8; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      addl(rtmp, offset);
+      pinsrb(dst, Address(base, rtmp), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  }
+}
+#endif // _LP64
+
+void C2_MacroAssembler::vgather8b(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base, Register rtmp, int vlen_enc) {
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+void C2_MacroAssembler::vgather8b_offset(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base,
+                                          Register offset, Register rtmp, int vlen_enc) {
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      movl(rtmp, Address(idx_base, i*4));
+      addl(rtmp, offset);
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      movl(rtmp, Address(idx_base, i*4));
+      addl(rtmp, offset);
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+void C2_MacroAssembler::vgather_subword(BasicType elem_ty, XMMRegister dst,  Register base, Register idx_base,
+                                        Register offset, Register mask, XMMRegister xtmp1, XMMRegister xtmp2, XMMRegister xtmp3,
+                                        Register rtmp, Register midx, Register length, int vector_len, int vlen_enc) {
+  assert(is_subword_type(elem_ty), "");
+  Label GATHER8_LOOP;
+  movl(length, vector_len);
+  vpxor(xtmp1, xtmp1, xtmp1, vlen_enc);
+  vpxor(dst, dst, dst, vlen_enc);
+  vallones(xtmp2, vlen_enc);
+  vpsubd(xtmp2, xtmp1, xtmp2 ,vlen_enc);
+  vpslld(xtmp2, xtmp2, 1, vlen_enc);
+  load_iota_indices(xtmp1, vector_len * type2aelembytes(elem_ty), T_INT);
+  bind(GATHER8_LOOP);
+    vpxor(xtmp3, xtmp3, xtmp3, vlen_enc);
+    if (offset == noreg) {
+      if (mask == noreg) {
+        vgather8b(elem_ty, xtmp3, base, idx_base, rtmp, vlen_enc);
+      } else {
+        LP64_ONLY(vgather8b_masked(elem_ty, xtmp3, base, idx_base, mask, midx, rtmp, vlen_enc));
+      }
+    } else {
+      if (mask == noreg) {
+        vgather8b_offset(elem_ty, xtmp3, base, idx_base, offset, rtmp, vlen_enc);
+      } else {
+        LP64_ONLY(vgather8b_masked_offset(elem_ty, xtmp3, base, idx_base, offset, mask, midx, rtmp, vlen_enc));
+      }
+    }
+    vpermd(xtmp3, xtmp1, xtmp3, vlen_enc == Assembler::AVX_512bit ? vlen_enc : Assembler::AVX_256bit);
+    vpsubd(xtmp1, xtmp1, xtmp2, vlen_enc);
+    vpor(dst, dst, xtmp3, vlen_enc);
```

Comment:
xtmp1 starts out as 0, 1,...
so vpermd will place the lower 64 bit from xtmp3 to lower 64 bit of dst
why vpsubd and not vpaddd? It looks to me that vpaddd is more intutive to understand.
if vpadd, xtmp1 will become 2,3 in next iteration 
so vpermd will place the lower 64 bit from xtmp3 to 127:64 of dst 
and so on so forth

Another point, for avx512 it looks to me that vpermd and vpor could be merged into one single instruction vpermd having dst as destination and merge bit set to true.

Reply:
Please ignore the last bit about avx512 vpermd merge as we are not using mask registers here.

Reply:
> xtmp1 starts out as 0, 1,... so vpermd will place the lower 64 bit from xtmp3 to lower 64 bit of dst why vpsubd and not vpaddd? It looks to me that vpaddd is more intutive to understand. if vpadd, xtmp1 will become 2,3 in next iteration so vpermd will place the lower 64 bit from xtmp3 to 127:64 of dst and so on so forth
> 
I have taken a different approach here based on progressive subtraction to get permute indices for each iteration.